server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java

```diff
@@ -396,7 +499,11 @@ public Object valueForDisplay(Object value) {
 
         @Override
         public Query existsQuery(QueryShardContext context) {
-            return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
+            if (hasDocValues()) {
+                return new DocValuesFieldExistsQuery(name());
+            } else {
+                return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
```

Comment:
This is not related to this pr but `KeyedJsonFieldType#existsQuery` should also use the `_field_names` field instead of a [prefix query](https://github.com/elastic/elasticsearch/blob/b2e0b996811ef80f020c5fa8690c7a1471f4f53c/server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java#L292) ?

Reply:
Thanks, I will give this some thought and follow up in a separate issue/ PR.